### DISCUSSION TITLE
Using Allegro for the first time(s)

### DISCUSSION THREAD

**Discussion by Timzoid** [asigalov61/Allegro-Music-Transformer#1] [Ideas]
I didn't do any controlled experiments, comparing LAMC to Allegro or Euterpe-X, measuring the time to generate simple continuations on the same intro, but it seems to work pretty much the same as far as the quality of the output.  I don't think Allegro is as fast as the "fast" version of Euterpe-X but I'm not sure. I have been using LAMC nearly every day, and used Euterpe-X only about 8 times. This was my second day using Allegro.

My tutorial on LAMC is up to over 1000 views, so maybe out of that, a hundred or so people have tried it, who may otherwise have not. That makes me feel good because it is a good substitute for MuseNet. 

 Since I have used MIDI software for years, and composed years before there were PCs, as well as being an early user of music software, I still see AI software like this for its collaborative potential, rather than its ability, or lack thereof, to just spit out a complete work that is any good. Even if AI software were to remain at this state, if the generation of continuations and transformations got faster, it's going to be a major part of the music scene.

Thanks for fixing the sequence generation tokens slider on LAMC, and I tried it today and am going to use the Composer Version more than I have, since for the last 30 pieces I've posted using your software to help create, I've been using the "Original Version" and mostly the "Inpaint" feature, but also the Simple Continuation now and then. I used the "Composer Version" for my first two pieces out of thirty. 

There are a lot of little things that I still miss about MuseNet -- the speed at which it could generate. Also, I had the ability to work on ten pieces at once, if I wanted to. Usually I worked on three at once. It was just efficient. If one piece turned out to be a flop, the other two might work out. 

In my tutorial for LAMC, which I made after less than a week of using the software, I warned people to save their Composition file after each block is added, because GitHub can disconnect and sometimes you can't get your file back. 

In the tutorial, I also mentioned how it was good for composers to save any simple continuations they like, ones they don't use to build the piece, because that is easy to do, and they are nice to have on hand. 

I think I mentioned before that my tutorial for MuseNet is up to over 31,000 views. It got less than 200 views while MuseNet was working, but ChatGPT had not yet launched. I've changed the title, including "MuseNet is dead" is the YouTube title, and people still watch it and think it is working because they go to the MuseNet page and they have that dummy version there, and no mention that the real version is discontinued. But I leave that tutorial up, because that's the best place to direct people to LAMC.

One suggestion I have, and it's probably not worth implementing on your existing AI music software, but when I was new to LAMC, I was confused in trying to upload a custom seed, that the file needed to be loaded into the file directory of the CoLab Notebook. I was just copying the path from my PC.

However, also on GitHub, an older AI program I only use occasionally, called "AI Forever Music Composer," that allows you to just point at the file from your computer that you want to upload as a custom seed and it does it, no extra copying the path name necessary. Also, the downloading of files you want to save is very convenient, as you don't have to go into the directory. The problem with that software is it is limited, although not bad for generating song ideas.

**Comment by asigalov61** [comment]
@Timzoid Thank you for your thoughts and feedback on Allegro and other stuff.

Allegro is as fast as Euterpe Small BUT because Allegro is using asynchronous encoding, so you get more notes per number of generated tokens. It is not a big difference but sometimes it is helpful.

Also, Allegro uses chords counters tokens which means it is just like MuseNet in terms of composition generation. Namely, Allegro will end the composition at some point if it is long enough just like MuseNet.

You are welcome about the sliders in LAMC. I will add the same to Euterpe shortly. I think it was a great suggestion and now it is much more useful.

Yes, MuseNet was great. The reason why it was so fast and nice was because they used heavily optimized code and Enterprise-grade equipment which is much faster than consumer stuff. Also MuseNet had additional things like model output critic and kick-ass interface. Plus it was free and web-based so no wonder it was so good :)

Now, RE: your Custom MIDI upload/path suggestion: There is a widget that can be attached there to upload MIDIs directly from user's computer, but it does not work too well in colab and it would make things more confusing. Otherwise, I will think about it and I will see what can be done. Maybe adding a note to the Custom MIDI section would clear things up and will be more easier to do anyway.

Anyway, thank you again for your feedback and suggestions. I really appreciate it and it helps me to improve my work :)

Sincerely,

Alex

**Comment by asigalov61** [comment]
@Timzoid Hey bro! :)

Check out my latest online version of Allegro Music Transformer :) 
https://huggingface.co/spaces/asigalov61/Allegro-Music-Transformer

I think you will like it and I would love to hear your thoughts and feedback! :)

Sincerely,

Alex

**Reply by Timzoid** [reply]
This is really good Alex. Now I can generate original seeds, even when my time on GitHub is out. And today, Sept 15, I notice you changed the interface, adding an "interrupt" button which is VERY helpful. By the  look of the pitch pattern, I can tell when something is not going to be useable, if it's like one line and I want polyphony, or sometimes just hammering away on one note.

A few minor things that I might point out in regard to terminology, especially for people looking at this for the first time --I have never seen an instrument listed like you have Drums "or not." The first time I saw that was in LAMC, and I stared at it trying to figure out what it meant. Just put, "Use Drums" or "Add Drums" and leave an empty box to the left, just as it is now. Axe the "or not." "Or not" is implied, a given. Also, on the subject of drums, you have the words "Drums Controls" there also. Not necessary, unless you're planning on adding some controls for drums. 

I understand how hard it is to approach these things like you're seeing it for the first time, as I am, let alone people who are new to this type of software. In fact, this Huggingface implementation of Allegro Music Transformer could be great for children, as well as my viewers who are not technical enough to understand CoLab, even with my full tutorial video. I haven't tried it on my phone, but it can probably work fine on that, too. 

I don't know what plans you have for adding more features to the Huggingface AMT, but for me, it would be great if it could do a continuation, and also, if the slider could be expanded, up to just over 1000 tokens, although 512 is not a bad size at all. It is a good seed size. For all I know, Huggingface puts limitations on these things. But the way it is today, with the slider in the middle as the default, that's a good idea for new users, who won't be discouraged that the sequence generation takes too long.

**Comment by asigalov61** [comment]
@Timzoid Thank you for your feedback as usual :)

I am glad you liked online demo. It is indeed great for generating seeds or short excerpts. I tried doing too with good results and convenience of MuseNet :)

I did what you asked and changed the drums button descriptions. I hope you like it now :)

RE: Continuations... I do not plan to add it because it is pretty complex and also due to it being too slow for that. 

Now, in regard to making it faster...I would love to do that but the prices are very high for that. Hugging Face charges min 0.6 cents per hour ($450/month) for a basic GPU. I can't afford that unfortunately :( If you can help with that by some chance, that would be great but I will understand if you can't.

Thank you again for trying it all out and thank you for your feedback/thoughts.

Alex

**Reply by fakerybakery** [reply]
Hi, have you looked into Hugging Face's Community GPU grant program?

**Reply by asigalov61** [reply]
@fakerybakery Yes, but it does not seem to be working atm...It would be great though if they could help with GPU :)

**Comment by asigalov61** [comment]
@Timzoid PS. Euterpe X is online too if you want to try it out or use it:

https://huggingface.co/spaces/asigalov61/Euterpe-X

Also, there is a similar project from which I took the code to make online demos. Try it out too. It plays ok and has most features and it is fast too (runs on GPU)

https://huggingface.co/spaces/skytnt/midi-composer

Alex

**Comment by asigalov61** [comment]
@Timzoid Hey, 

Just wanted to let you know that I speed up the generation x2 with linear batches speed scaling. Check it out and let me know if you like it now :)

Alex

**Reply by fakerybakery** [reply]
Hi @asigalov61
Thanks for sharing this update! Is the same method available for Euterpe-X?
Thanks in advance!

**Comment by asigalov61** [comment]
@fakerybakery Yes, I already put it in Euterpe-X Composer code/colab and I will add it in other Euterpe X colabs shortly. I will also add it to my other implementations as well shortly :)

Alex
